docs/birthday-kata.md

```diff
@@ -0,0 +1,890 @@
+
+As an aside, let's avoid modeling failures simply because it will clutter this exercise a bit. In real life, many of our effects would have type signatures with `Either` so we could model possible failures, but let's just naively pretend everything always works.
+
+So let's do some more scaffolding!
+```unison
+sendBirthdayEmails : '[Message]
+sendBirthdayEmails = 
+  today' = today
+  employees = fetchAllByBirthday (Date.day today', Date.month today')
+  messages = map (Employee.toBirthdayMessage) employees
+  map (send) messages
+
+Employee.toBirthdayMessage : Employee -> Message
+Employee.toBirthdayMessage _ = todo "
+```
+So we will grab the date, use that fetch all the employees with birthdays today, map those employees to messages, and then send out (and return) those messages.
+
+Uh oh! The compiler is mad!
+```ucm
+The expression in red needs the {Calendar} ability, but this location does not have access to any abilities.
+  
+    136 |   today' = today)
+```
+
+Ah, turns out that any function with an ability in its signature can only be called by another function with that ability available in its signature. So we need to add all our abilities to the function signature. So let's just go ahead and add all three.
+
+```unison
+sendBirthdayEmails : '{MessageService, EmployeeRepository, Calendar} [Message]
+```
+
+But we are stil getting the same error :(
+
+Well, it turns out that these effects can actually only be run in the context of a handler. Until we wire that up, we can instead just defer the computation (so all affects become deferred).
+
+```unison
+sendBirthdayEmails : '[Message]
+sendBirthdayEmails =  'let
+  today' = today
+  employees = fetchAllByBirthday (Date.day today', Date.month today')
+  messages = map (Employee.toBirthdayMessage) employees
+  map (send) messages
+
+Employee.toBirthdayMessage : Employee -> Message
+Employee.toBirthdayMessage _ = todo "toBirthdayMessage"
+```
+
+And the compiler seems happy so far! Note that we just deferred the entire computation by putting the `'let` at the beginning.
+
+Let's move on applying TDD in order to implement the remaining code.
+
+## Implementation!
+Let's write a first failing test!
+```unison
+test> sendBirthdayEmails.tests.noBirthdaysToday = 
+  check let
+    actual = !sendBirthdayEmails
+    expected = []
+    if expected === actual
+    then true
+    else bug (exected, actual)
+```
+
+```ucm
+  The expression in red  needs these abilities: {Calendar, EmployeeRepository, 𝕖457}, but this location does not have access to any abilities.
+  
+     70 |     actual = !sendBirthdayEmails
+```
+
+Well we can't defer this call any more, this is the boundary of the system! So it's time to implement handlers for our abilities!
+
+```unison
+test> sendBirthdayEmails.tests.noBirthdaysToday = 
+  check let
+    actual = 
+      handle 
+        handle 
+          handle 
+            !sendBirthdayEmails 
```

Comment:
I'm gonna think on this structure overnight 🤔 but I think this is combining two concerns, it's creating a conglomerate mock handler for our core business logic, and also doing the test assertion. One thing we might do is separate the test assertion from the handler for better code re-use.  

```
use Universal 
testBusinessLogicHandler : '{MessageService, EmployeeRepository, Calendar} a -> a  
testBusinessLogicHandler myBusinessLogic = 
  handle 
    handle 
      handle 
        !myBusinessLogic 
      with (Calendar.handler.mock (Date (Day 1) January (Year 1991)))
    with (EmployeeRepository.handler.mock [])
  with MessageService.handler.mock

test> sendBirthdayEmails.tests.noBirthdaysToday = 
  check let
    actual = testBusinessLogicHandler sendBirthdayEmails
    expected = []
    if expected === actual
    then true
    else bug (expected, actual)
```

This might be better for future tests because you might imagine other tests using this test fixture, or a more general one which would let the caller of the mock handler provide expectations: `testBusinessLogicHandler2 : Date -> [Employee] -> '{MessageService, EmployeeRepository, Calendar} a -> a `

Reply:
Hah, silly me I should have read a bit farther down where you do exactly this.